src/release.rs

```diff
@@ -0,0 +1,148 @@
+            "tar" => Ok(DeployMethod::Tar),
+            "skip" => Ok(DeployMethod::Skip),
+            _ => Err(()),
+        }
+    }
+}
+
+pub struct BuildConfig {
+    deploy_method: DeployMethod,
+    skip_build: bool,
+    debug_build: bool,
+    _build_path: PathBuf,
+    solana_root_path: PathBuf,
+}
+
+impl BuildConfig {
+    pub fn new(
+        deploy_method: &str,
+        skip_build: bool,
+        debug_build: bool,
+        solana_root_path: &PathBuf,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        let deploy_method = deploy_method
+            .parse::<DeployMethod>()
+            .map_err(|_| "Failed to parse deploy_method".to_string())?;
+
+        let build_path = match deploy_method {
+            DeployMethod::Local => solana_root_path.join("farf/bin"),
+            DeployMethod::Tar => solana_root_path.join("solana-release/bin"),
+            DeployMethod::Skip => solana_root_path.join("farf/bin"),
+        };
+
+        Ok(BuildConfig {
+            deploy_method,
+            skip_build,
+            debug_build,
+            _build_path: build_path,
+            solana_root_path: solana_root_path.clone(),
+        })
+    }
+
+    pub async fn prepare(&self) -> Result<(), Box<dyn Error>> {
+        match self.deploy_method {
+            DeployMethod::Tar => {
+                return Err(boxed_error!("Tar deploy method not implemented yet."));
+            }
+            DeployMethod::Local => {
+                self.setup_local_deploy()?;
+            }
+            DeployMethod::Skip => {
+                return Err(boxed_error!("Skip deploy method not implemented yet."));
+            }
+        }
+        info!("Completed Prepare Deploy");
+        Ok(())
+    }
+
+    fn setup_local_deploy(&self) -> Result<(), Box<dyn Error>> {
+        if !self.skip_build {
+            self.build()?;
+        } else {
+            info!("Build skipped due to --skip-build");
+        }
+        Ok(())
+    }
+
+    fn build(&self) -> Result<(), Box<dyn Error>> {
+        let start_time = Instant::now();
+        let build_variant = if self.debug_build { "--debug" } else { "" };
+
+        let install_directory = self.solana_root_path.join("farf");
+        let install_script = self.solana_root_path.join("scripts/cargo-install-all.sh");
+        match std::process::Command::new(install_script)
+            .arg(install_directory)
+            .arg(build_variant)
+            .arg("--validator-only")
+            .status()
+        {
+            Ok(result) => {
+                if result.success() {
+                    info!("Successfully built validator")
+                } else {
+                    return Err(boxed_error!("Failed to build validator"));
+                }
+            }
+            Err(err) => return Err(Box::new(err)),
+        }
+
+        let solana_repo = Repository::open(self.solana_root_path.as_path())?;
+        let commit = solana_repo.revparse_single("HEAD")?.id();
```

Comment:
it would be wonderful if we can specify a commit hash :trollface:

Reply:
how would you see this working? a commit hash for the upstream agave repo? What about forked repos? we could do `--deploy-method hash --hash <commit-hash>`. Would this then pull from some Agave monorepo hash? That seems easy enough. Or are you also thinking some forked repo hash. so something like: `--deploy-method hash --hash <commit-hash> --repo <forked-agave-repo>`?

Reply:
I can do this. but i would probably put in another PR

Reply:
looks like we will only use HEAD here 🤔 it means I need to cd into the path and checkout to the correct one.
tbh, I guess I more like to have a feature `branch_or_hash`. we can deploy two different cluster quickly to compare their performance without any other actions.

but yeah. we can do it in the future! it is a nice to have!

Reply:
feel free to merge. generally I think this PR is a good start!

Reply:
ya that is true. it will use the HEAD of whatever branch you are on in your local `solana` monorepo you have as passed in via `--local-path <path-to-solana-repo>`